compiler/rustc_trait_selection/src/traits/fulfill.rs

```diff
@@ -210,6 +210,20 @@ impl<'a, 'tcx> ObligationProcessor for FulfillProcessor<'a, 'tcx> {
     type Error = FulfillmentErrorCode<'tcx>;
     type OUT = Outcome<Self::Obligation, Self::Error>;
 
+    #[inline]
+    fn skippable_obligations<'b>(
+        &'b self,
+        it: impl Iterator<Item = &'b Self::Obligation>,
+    ) -> usize {
+        let is_unchanged = self.selcx.infcx.is_ty_infer_var_unchanged();
+
+        it.take_while(|o| match o.stalled_on.as_slice() {
+            [o] => is_unchanged(*o),
```

Comment:
This will pessimize cases where `stalled_on` frequently has 2 elements, such as `mime-0.3.16`,  `wast-54.0.0`, and `inflate-0.4.5`. Hopefully not too much.

(If you want to measure these locally, download with `target/release/collector download crate $NAME $VERSION`.)

Reply:
Also seems to benefit. Let's see if perf agrees.

```
mime-0.3.16	check	full	-0.39% 	0.20%	1.97x
```